Tree/README.markdown

```diff
@@ -151,6 +153,149 @@ tree.search("chai")     // returns the "chai" node
 tree.search("bubbly")   // nil
 ```
 
+## Solution 2 (using an enum)
+
+The first solution uses a class to represent the Tree datastructure. Classes are passed by reference in Swift, meaning that if you copy your object and modify it, that will also affect both of your references, because they are referencing the exact same place in data:
+
+```swift
+let sameTree = tree
+sameTree.value = "drinks"
+
+sameTree.value   // "drinks"
+tree.value       // also "drinks"
+```
+
+While this can be desired in some cases, in other cases you want to have something, that is passed by value, meaning that it is copied every time you hand it to a function, assign it to a variable, etc. To implement this, we will use and enum:
+
+```swift
+enum TreeNode<T> {
+  case Leaf(value: T)
+  indirect case Node(value: T, children: [TreeNode])
+  
+  func add(child: TreeNode) -> TreeNode {
+    switch self {
+    case .Leaf(let ownValue):
+      return .Node(value: ownValue, children: [child])
+    case .Node(let ownValue, let ownChildren):
+      return .Node(value: ownValue, children: ownChildren+[child])
+    }
+  }
+}
+```
+
+The enum has two cases:
+
+- `Leaf` for a leaf node that has no children
+- `Node` for an node that has one or more children. It is marked indirect to be able to hold `TreeNode`s. Without `indirect` an enum can't hold instances of itself
+
+> **Note:** As the enum-implementation for a [binary search tree](../Binary Search Tree/) this implementation does not hold a reference to the parent node.
+
+As in the implementation using a class, let's make this tree printable by conforming to the `CustomStringConvertible` protocol:
+```swift
+extension TreeNode: CustomStringConvertible {
+  var description: String {
+    switch self {
+    case .Leaf(let value):
+      return "\(value)"
+    case .Node(let value, let children):
+      return "\(value) {" + children.map { $0.description }.joined(separator: " ,") + "} "
+    }
+  }
+}
+```
+
+And again, to see this in action in a playground:
+```swift
+var tree = TreeNode.Leaf(value: "beverages")
+
+var hotNode = TreeNode.Leaf(value: "hot")
+var coldNode = TreeNode.Leaf(value: "cold")
+
+var teaNode = TreeNode.Leaf(value: "tea")
+var coffeeNode = TreeNode.Leaf(value: "coffee")
+var chocolateNode = TreeNode.Leaf(value: "cocoa")
+
+var blackTeaNode = TreeNode.Leaf(value: "black")
+var greenTeaNode = TreeNode.Leaf(value: "green")
+var chaiTeaNode = TreeNode.Leaf(value: "chai")
+
+var sodaNode = TreeNode.Leaf(value: "soda")
+var milkNode = TreeNode.Leaf(value: "milk")
+
+var gingerAleNode = TreeNode.Leaf(value: "ginger ale")
+var bitterLemonNode = TreeNode.Leaf(value: "bitter lemon")
+
+sodaNode = sodaNode.add(gingerAleNode)
```

Comment:
This part doesn't compile in Swift 3, since your `add` method has an explicit argument.